Support correlated subqueries in WHERE clause

novikd opened the pull request:
### Changelog category (leave one):

- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Support correlated subqueries as an argument of `EXISTS` expression in the WHERE clause. Closes #72459.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

novikd committed:
WIP: Correlated subqueries

Algunenano commented on src/Core/Settings.cpp:
```diff
@@ -4569,6 +4569,10 @@ Connect timeout in seconds. Now supported only for MySQL
+    \
+    DECLARE(Bool, allow_experimental_correlated_subqueries, false, R"(
+Allow to execute correlated subqueries.
+)", 0) \
```
Please remember to tag it as `EXPERIMENTAL` in the flags.

qoega commented:
Now CI should start if you will update your branch

novikd commented:
@qoega it didn't help

qoega commented:
I propose to change to ordinary feature now until Max is back

novikd committed:
Remove redundant logging

novikd committed:
Merge remote-tracking branch 'origin/master' into analyzer-scalar-correlated-subqueries

novikd committed:
Build correlated EXISTS

novikd committed:
Add lost file

novikd committed:
Analyze expressions with correlated subqueries and build plan to decorrelate

novikd committed:
Merge remote-tracking branch 'origin/master' into analyzer-scalar-correlated-subqueries

novikd committed:
Small progress

novikd committed:
Create PLACEHOLDER when correlated column is used

novikd committed:
WIP: Build decorrelated plan

novikd committed:
Merge remote-tracking branch 'origin/master' into analyzer-scalar-correlated-subqueries

novikd committed:
Build decorrelated plan

novikd committed:
Simplify code a little bit

novikd committed:
Implement ReadFromMergeTree::clone()

novikd committed:
Simplify headers

novikd committed:
Materialize EXISTS result column

novikd committed:
Read columns used in subquery

novikd committed:
Do no preserve column after decorrelate

novikd committed:
Fix crash

novikd committed:
Fix 02995_new_settings_history

novikd committed:
Fix EXISTS to IN conversion

novikd committed:
Update reference file

novikd committed:
Merge remote-tracking branch 'origin/master' into analyzer-scalar-correlated-subqueries

novikd committed:
Fix build

novikd committed:
Fix build v2

novikd committed:
Fix cross replication

Fix TableNode::updateTreeHashImpl in case of cross replication.
Hash of TableNode can be calculated during ColumnNode hash
calculation because it's column source

novikd committed:
Merge remote-tracking branch 'origin/master' into analyzer-scalar-correlated-subqueries

novikd committed:
Fix sanitizer error

novikd committed:
Run tests only with enabled analyzer

novikd committed:
Fix ValidationChecker

novikd committed:
Fix build

novikd committed:
Clean up code

novikd committed:
Build plan for correlated subquery with UNION

novikd committed:
Fix another sanitizer assert

novikd committed:
Merge remote-tracking branch 'origin/master' into analyzer-scalar-correlated-subqueries

novikd committed:
Yet another build fix

novikd committed:
Clean up code

novikd committed:
Add comments for buildQueryPlanForCorrelatedSubquery

novikd committed:
Add a test for #72459

novikd committed:
Fix typo

novikd committed:
Merge remote-tracking branch 'origin/master' into analyzer-scalar-correlated-subqueries

novikd committed:
Improve comments

novikd committed:
Support UnionStep in correlated plan

novikd committed:
Improve test

novikd committed:
Fix used JOIN strictness

KochetovNicolai reviewed (commented):
Overall looks good.
Still, I did not read all the details of PlannerCorrelatedSubqueries.cpp yet.

novikd committed:
Some review fixes

novikd committed:
Make remove unused projections respect correlated columns

novikd committed:
Fix error code for correlated subqueries in IN expression

novikd committed:
Run test only with analyzer

novikd committed:
Fix QueryNode::clone and UnionNode::clone

novikd committed:
Fix build

KochetovNicolai commented:
Fuzzer debug: `SELECT json FROM test ORDER BY ALL ASC NULLS LAST LIMIT 1025, 65535`. Not related.

KochetovNicolai commented:
Fuzzer tsan: `SELECT number FROM mem2 AS tbl INNER JOIN (SELECT number FROM numbers(1) WHERE tbl.number >= number) AS alias4 ON alias4.number = number settings allow_experimental_correlated_subqueries=1
`. Related.

KochetovNicolai committed:
Add one more check.
